source/esbuild-plugin.civet

```diff
@@ -13,34 +15,115 @@ esbuild.build({
+const { solidPlugin } = require('esbuild-plugin-solid');
+import civetPlugin from '@danielx/civet/esbuild-plugin'
+
+esbuild.build({
+  ...,
+  plugins: [
+    civetPlugin(
+      next: solidPlugin()
+    )
+  ]
+}).catch(() => process.exit(1))
+```
 */
 
-{ readFile } := require 'fs/promises'
+import type { Plugin, OnLoadArgs, OnLoadResult } from 'esbuild'
+
+interface Options {
+  filter?: RegExp
+  inlineMap?: boolean
+  js?: boolean
+  next?: Plugin
+}
+
+{ readFile, writeFile, mkdir } := require 'fs/promises'
 path := require 'path'
 
 // NOTE: this references the built version of the module, not the source
 { compile } := require "../dist/main.js"
 
-module.exports = {
-  name: 'civet'
-  setup(build)
-    build.onLoad {
-      filter: /\.civet$/
-    }, (args) ->
-      try
-        source := await readFile args.path, 'utf8'
-        filename := path.relative(process.cwd(), args.path)
-        compiled := compile source, {
-          filename
-          inlineMap: true
-          js: true
-        }
-
-        return
-          contents: compiled
-      catch e
-        return
-          errors: [{
-            text: e.message
-          }]
-}
+// NOTE: this function is named civet so esbuild gets "civet" as the name of the plugin
+civet := (options: Options = {}): Plugin ->
+  {
+    filter=/\.civet$/
+    inlineMap=true
+    js=true
+    next
+  } := options
+
+  let nextTransform: (args: OnLoadArgs) => OnLoadResult | Promise<OnLoadResult | null | undefined> | null | undefined
+  let tmpPath: null | string
+  let dirCreated = false
+
+  if next
+    { tmpdir } := require 'os'
+    tmpPath = path.join tmpdir(), `civet-${process.pid}`
+
+    next.setup {
+      onEnd()
+        ;
+      onStart()
+        ;
+      resolve()
+        ;
+      onResolve()
+        ;
+      initialOptions()
+        ;
+      esbuild()
+        esbuild
+      onLoad(_, handler)
+        nextTransform = handler
+    }
+
+  return {
+    name: "civet"
+    setup(build)
+      build.onLoad { filter }, (args) ->
+        try
+          source := await readFile args.path, 'utf8'
+          filename := path.relative(process.cwd(), args.path)
+          compiled := compile source, {
+            filename
+            inlineMap
+            js
+          }
+
+          if next
+            if !dirCreated
+              await mkdir tmpPath
+              dirCreated = true
+
+            outputFileName := filename + js ? '.jsx' : '.tsx'
+
+            // I'd prefer not to use temp files but I can't find a way to pass a stream to fs.readFile which is what
+            // most esbuild plugins use
+            await writeFile path.join(tmpPath, outputFileName), compiled
+            return await nextTransform {
+              ...args
+              path: path.join(tmpPath, filename)
+            }
+
+          return
+            contents: compiled
+        catch e
+          return
+            errors: [{
+              text: e.message
+            }]
+  }
```

Comment:
Is there a "cleanup" mechanism in esbuild plugins, so we could delete the temporary directory?  Not a big deal but it'd be nice.